packages/headless-react/src/search-engine.test.tsx

```diff
@@ -0,0 +1,145 @@
+import {getSampleSearchEngineConfiguration} from '@coveo/headless';
+import {InferCSRState, defineResultList} from '@coveo/headless/ssr';
+import '@testing-library/jest-dom';
+import {render} from '@testing-library/react';
+import {useEffect, useState} from 'react';
+import {defineSearchEngine} from './search-engine';
+
+// TODO(KIT-2692): Disabled until JSDOM env error can be fixed
+//  Error: "TypeError: Cannot read properties of undefined (reading 'html')"
+describe.skip('Headless react SSR utils', () => {
+  let errorSpy: jest.SpyInstance;
+  const sampleConfig = {
+    ...getSampleSearchEngineConfiguration(),
+    analytics: {enabled: false}, // TODO: KIT-2585 Remove after analytics SSR support is added
+  };
+
+  beforeEach(() => {
+    errorSpy = jest.spyOn(global.console, 'error');
+  });
+
+  afterEach(() => {
+    errorSpy.mockClear();
+  });
+
+  test('defines react search engine', () => {
+    const {
+      fetchInitialState,
+      hydrateInitialState,
+      build,
+      useEngine,
+      controllers,
+      SSRStateProvider,
+      CSRProvider,
+      ...rest
+    } = defineSearchEngine({
+      configuration: sampleConfig,
+    });
+
+    [
+      fetchInitialState,
+      hydrateInitialState,
+      build,
+      useEngine,
+      SSRStateProvider,
+      CSRProvider,
+    ].forEach((returnValue) => expect(typeof returnValue).toBe('function'));
+
+    expect(controllers).toEqual({});
+    expect(rest).toEqual({}); // No other return values
+  });
+
+  test('creates a hook based on given controller', () => {
```

Comment:
I think we should also add some tests to verify the hooks work correctly.

E.g.:
1. When using the SSRStateProvider, returns a `state` but no `methods`.
2. When using the CSRProvider
   a. Returns a `state` and `methods`.
   b. Subscribes to the controller and updates the `state` on change.
   c. Reacts correctly when a method is called (e.g., `searchBox.submit()` should update the state).
3. When switching from SSRStateProvider to CSRProvider
   a. Should cause a re-render / update the state.